fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/api/attachment/v1/Attachment.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.attachment.v1;
+
+import java.util.function.Supplier;
+
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.util.Identifier;
+
+/**
+ * An attachment allows "attaching" arbitrary data to various game objects (entities, block entities, worlds and chunks at the moment).
+ * Use the methods provided in {@link AttachmentRegistry} to create and register attachments.
+ *
+ * <p>Attachments can optionally be made to persist between restarts using {@link AttachmentSerializer}s, and
+ * can optionally be automatically synced between server and client.</p>
+ *
+ * @param <A> type of the attached data
+ */
+public interface Attachment<A> {
```

Comment:
I think `Attachment` should be renamed to `AttachmentType` (as in the original PR) since it is really a type or key and not an instance. This is also similar to what vanilla does with `IntProvider`, `TreeTrunkProvider`, `BrainMemoryType`, etc.